src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/ContractHelper.cs

```diff
@@ -0,0 +1,159 @@
+                        }
+                    }
+                    if (eventArgs.Unwind)
+                    {
+                        // unwind
+                        if (innerException == null) { innerException = eventArgs.thrownDuringHandler; }
+                        throw new ContractException(failureKind, displayMessage, userMessage, conditionText, innerException);
+                    }
+                }
+            }
+            finally
+            {
+                if (eventArgs != null && eventArgs.Handled)
+                {
+                    returnValue = null; // handled
+                }
+                else
+                {
+                    returnValue = displayMessage;
+                }
+            }
+            return returnValue;
+        }
+
+        private static string GetResourceNameForFailure(ContractFailureKind failureKind)
+        {
+            string resourceName = null;
+            switch (failureKind)
+            {
+                case ContractFailureKind.Assert:
+                    resourceName = "AssertionFailed";
+                    break;
+
+                case ContractFailureKind.Assume:
+                    resourceName = "AssumptionFailed";
+                    break;
+
+                case ContractFailureKind.Precondition:
+                    resourceName = "PreconditionFailed";
+                    break;
+
+                case ContractFailureKind.Postcondition:
+                    resourceName = "PostconditionFailed";
+                    break;
+
+                case ContractFailureKind.Invariant:
+                    resourceName = "InvariantFailed";
+                    break;
+
+                case ContractFailureKind.PostconditionOnException:
+                    resourceName = "PostconditionOnExceptionFailed";
+                    break;
+
+                default:
+                    Debug.Fail("Unreachable code");
+                    resourceName = "AssumptionFailed";
+                    break;
+            }
+            return resourceName;
+        }
+
+        private static string GetDisplayMessage(ContractFailureKind failureKind, string userMessage, string conditionText)
+        {
+            string failureMessage;
+            // Well-formatted English messages will take one of four forms.  A sentence ending in
+            // either a period or a colon, the condition string, then the message tacked 
+            // on to the end with two spaces in front.
+            // Note that both the conditionText and userMessage may be null.  Also, 
+            // on Silverlight we may not be able to look up a friendly string for the
+            // error message.  Let's leverage Silverlight's default error message there. 
+            if (!string.IsNullOrEmpty(conditionText))
+            {
+                string resourceName = GetResourceNameForFailure(failureKind); 
+                resourceName += "_Cnd";
```

Comment:
I would keep the CoreRT variant that calls the `SR.` methods. That pattern is friendly to resource string tree shaking and similar resource string optimizations. Constructing the resource string dynamically is not.

Reply:
I certainly prefer the CoreRT version, but I saw that `GetResourceNameForFailure` is actually called twice from different places. Let me check if the value is used in the other place. If not I will follow your suggestion and switch this to the CoreRT code.